sdk/rust/oak/src/grpc/server.rs

```diff
@@ -16,27 +16,68 @@
+    }
+}
+
+// TODO(#389): Automatically generate this code.
+impl crate::io::Decodable for GrpcServerInit {
+    fn decode(message: &crate::io::Message) -> Result<Self, crate::OakError> {
+        if !message.bytes.is_empty() {
+            panic!(
+                "incorrect number of bytes received: {} (expected: 0)",
+                message.bytes.len()
+            );
+        }
+        if message.handles.len() != 1 {
+            panic!(
+                "incorrect number of handles received: {} (expected: 1)",
+                message.handles.len()
+            );
+        }
+        Ok(GrpcServerInit {
+            invocation_sender: crate::io::Sender::new(crate::WriteHandle {
+                handle: message.handles[0],
+            }),
+        })
+    }
+}
 
 /// Initializes a gRPC server pseudo-Node listening on the provided address.
 ///
-/// Returns a [`ReadHandle`] to read invocations from.
-///
-/// [`ReadHandle`]: crate::ReadHandle
-pub fn init(address: &str) -> std::result::Result<ReadHandle, OakStatus> {
-    // Create a channel and pass the read half to a new gRPC pseudo-Node.
-    let (write_handle, read_handle) = crate::channel_create().expect("Couldn't create a channel");
+/// Returns a [`Receiver`] to read gRPC [`Invocation`]s from.
+pub fn init(address: &str) -> Result<Receiver<Invocation>, OakStatus> {
+    // Create a channel and pass the read half to a new gRPC server pseudo-Node.
+    let (init_sender, init_receiver) = crate::io::channel_create::<GrpcServerInit>()
+        .expect("Couldn't create init channel to gRPC server pseudo-node");
     let config = crate::node_config::grpc_server(address);
-    crate::node_create(&config, read_handle)?;
-    crate::channel_close(read_handle.handle).expect("Couldn't close a channel");
+    crate::node_create(&config, init_receiver.handle)?;
+    init_receiver
+        .close()
+        .expect("Couldn't close init channel to gRPC server pseudo-node");
 
     // Create a separate channel for receiving invocations and pass it to a gRPC pseudo-Node.
-    let (invocation_write_handle, invocation_read_handle) =
-        crate::channel_create().expect("Couldn't create a channel");
-    crate::channel_write(write_handle, &[], &[invocation_write_handle.handle])
-        .expect("Couldn't write to a channel");
-    crate::channel_close(write_handle.handle).expect("Couldn't close a channel");
-    crate::channel_close(invocation_write_handle.handle).expect("Couldn't close a channel");
-
-    Ok(invocation_read_handle)
+    let (invocation_sender, invocation_receiver) =
+        crate::io::channel_create::<Invocation>().expect("Couldn't create gRPC invocation channel");
+
+    let grpc_server_init = GrpcServerInit { invocation_sender };
+    init_sender
+        .send(&grpc_server_init)
+        .expect("Could not send init message to gRPC server pseudo-node");
+    init_sender
+        .close()
+        .expect("Couldn't close init message channel to gRPC server pseudo-node");
```

Comment:
I think this induces a handle leak &ndash; the local Node's copy of the `invocation_sender` handle is not closed after it has been sent to the gRPC server p-N, and it is not saved anywhere so the handle value is lost.

Reply:
Raised as #1358.